music21/stream/base.py

```diff
@@ -8988,46 +9013,58 @@ def bestMatch(target, divisors):
-                found.append((error, match, signedErrorInner))  # reverse match, error for sorting
+                # reverse match, error for sorting and add divisor expressed as qL
+                found.append((error, match, signedErrorInner, 1 / div))
             # get first, and leave out the error
             bestMatchTuple = sorted(found)[0]
             return bestMatchTuple
 
         # if we have a min of 0.25 (sixteenth)
         # quarterLengthMin = quarterLengthDivisors[0]
 
-        # TODO: Use new filters...
         if inPlace is False:
             returnStream = self.coreCopyAsDerivation('quantize')
-
         else:
             returnStream = self
 
         useStreams = [returnStream]
         if recurse is True:
-            for obj in returnStream.recurse():
-                if 'Stream' in obj.classes:
-                    useStreams.append(obj)
+            useStreams = returnStream.recurse(streamsOnly=True, includeSelf=True)
 
         for useStream in useStreams:
-            for e in useStream._elements:
+            for i, e in enumerate(useStream._elements):
                 if processOffsets:
                     o = useStream.elementOffset(e)
                     sign = 1
                     if o < 0:
                         sign = -1
                         o = -1 * o
-                    unused_error, oNew, signedError = bestMatch(float(o), quarterLengthDivisors)
+                    unused_error, oNew, signedError, unused_inverse_divisor = bestMatch(
+                        float(o), quarterLengthDivisors)
                     useStream.coreSetElementOffset(e, oNew * sign)
                     if (hasattr(e, 'editorial')
                             and signedError != 0):
                         e.editorial.offsetQuantizationError = signedError * sign
                 if processDurations:
                     ql = e.duration.quarterLength
                     ql = max(ql, 0)  # negative ql possible in buggy MIDI files?
-                    unused_error, qlNew, signedError = bestMatch(
+                    unused_error, qlNew, signedError, unused_inverse_divisor = bestMatch(
                         float(ql), quarterLengthDivisors)
+                    # Check that any gaps from this quantized duration to the next onset
+                    # are at least as large as the smallest quantization unit (largest divisor)
+                    # If not, then re-quantize this duration with the divisor
+                    # that will be used to quantize the next element's offset
+                    if processOffsets and i + 1 < len(useStream._elements):
+                        next_element = useStream._elements[i + 1]
+                        next_offset = useStream.elementOffset(next_element)
+                        look_ahead_result = bestMatch(float(next_offset), quarterLengthDivisors)
+                        next_offset = look_ahead_result[1]
+                        next_divisor = 1 / look_ahead_result[3]
```

Comment:
Because of this[1] , [3], I'd prefer that bestMatch returned a NamedTuple, so that (as above), we could avoid all the "unused" and see where `inverseDivisor` is actually being used.  It does seem though that the only time it is used, it is immediately flipped back, so should it just be stored as divisor rather than inverse?

Reply:
I needed it to sort with the larger divisor/smaller resulting QL _first_, so a little lazy/hacky, but could do with a code comment.

Reply:
nope... wrong... we already have `match` as 1/divisor and that's early enough in the sort, so I can ditch the fourth component and just take 1/match when I need divisor